# Individual homework project 2

You will apply statistical testing and correlation analysis to the ShopNow dataset.

For each question:

1. Run the appropriate test.
2. Report the test statistic and p-value (if applicable).
3. Interpret the result in plain English.
4. Explain what it means for the business.

In [ ]:
# Import modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [ ]:
# Import datasets
customers = pd.read_csv("customers.csv")
transactions = pd.read_csv("transactions.csv")
sessions = pd.read_csv("web_sessions.csv")
campaigns = pd.read_csv("marketing_campaigns.csv")

merged = transactions.merge(customers, on="customer_id")
sessions_merged = sessions.merge(customers, on="customer_id")

### Specific questions

Q1. Do customers from the UK spend more per transaction than customers from Germany?

In [ ]:
uk = merged[merged["country"] == "UK"]["purchase_value"]
germany = merged[merged["country"] == "Germany"]["purchase_value"]

t_stat, p_value = stats.ttest_ind(uk, germany)

print(f"UK average: ${uk.mean():.2f}")
print(f"Germany average: ${germany.mean():.2f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: The difference IS statistically significant")
else:
    print("Result: The difference is NOT statistically significant")

UK average: $75.63
Germany average: $75.76
T-statistic: -0.1120
P-value: 0.9108
Result: The difference is NOT statistically significant


Q1 Interpretation:
UK customers average 75.63 and Germany 75.76, a difference of only 13 cents.
P-value: 0.9108 not significant.

The difference is completely random. The business should not treat UK and German customers differently in terms of pricing or marketing spend.

Q2. Do discounted purchases differ in average value from non-discounted purchases?

In [ ]:
discounted = merged[merged["discount_used"] == 1]["purchase_value"]
not_discounted = merged[merged["discount_used"] == 0]["purchase_value"]

t_stat, p_value = stats.ttest_ind(discounted, not_discounted)

print(f"Discounted average: ${discounted.mean():.2f}")
print(f"Non-discounted average: ${not_discounted.mean():.2f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: The difference IS statistically significant")
else:
    print("Result: The difference is NOT statistically significant")

Discounted average: $64.40
Non-discounted average: $80.76
T-statistic: -27.1597
P-value: 0.0000
Result: The difference IS statistically significant


Q2 Interpretation:
Discounted purchases average 64.40 vs 80.76 for full price,a $16.37 difference.
P-value: 0.0000  significant.

This is the strongest finding so far. Discount customers consistently spend less, and this is 100% real, not random. The business should reconsider its discount strategy as it is reducing revenue per transaction.

Q3. Do male and female customers differ in average purchase value?

In [ ]:
male = merged[merged["gender"] == "Male"]["purchase_value"]
female = merged[merged["gender"] == "Female"]["purchase_value"]

t_stat, p_value = stats.ttest_ind(male, female)

print(f"Male average: ${male.mean():.2f}")
print(f"Female average: ${female.mean():.2f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: The difference IS statistically significant")
else:
    print("Result: The difference is NOT statistically significant")

Male average: $76.04
Female average: $75.31
T-statistic: 1.0828
P-value: 0.2789
Result: The difference is NOT statistically significant


Q3 Interpretation:
Males average 76.04 and females 75.31  the difference is only 73 cents.
P-value: 0.2789 not significant.

Gender does not influence spending amount. Creating gender-based pricing or promotions would not be justified based on this data.

Q4. Do customers acquired via email spend more than those from social channels?

In [ ]:
email = merged[merged["acquisition_channel"] == "email"]["purchase_value"]
social = merged[merged["acquisition_channel"] == "social"]["purchase_value"]

t_stat, p_value = stats.ttest_ind(email, social)

print(f"Email average: ${email.mean():.2f}")
print(f"Social average: ${social.mean():.2f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: The difference IS statistically significant")
else:
    print("Result: The difference is NOT statistically significant")

Email average: $75.43
Social average: $76.04
T-statistic: -0.6980
P-value: 0.4852
Result: The difference is NOT statistically significant


Q4 Interpretation:
Email customers average 75.43 and social customers 76.04, a difference of 61 cents.
P-value: 0.4852 not significant.

The channel that brought a customer in does not affect how much they spend. Email and social media customers are equally valuable in terms of transaction value.

Q5. Is device type associated with conversion?

In [ ]:
contingency_table = pd.crosstab(sessions["device_type"], sessions["converted"])
print(contingency_table)

chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

print(f"\nChi-square statistic: {chi2:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: Device type IS associated with conversion")
else:
    print("Result: Device type is NOT associated with conversion")

converted        0     1
device_type             
desktop      20020  3068
mobile       20213  3042
tablet       20244  3013

Chi-square statistic: 1.1502
P-value: 0.5626
Result: Device type is NOT associated with conversion


Q5 Interpretation:
Desktop: 3068 conversions, Mobile: 3042, Tablet: 3013.
P-value: 0.5626 not significant.

Device type has no association with whether someone buys or not. All devices convert at roughly the same rate (~13%). This is good news, it means the website works equally well on mobile, tablet and desktop.

Q6. Is acquisition channel associated with conversion?

In [ ]:
sessions_merged2 = sessions.merge(customers, on="customer_id")

contingency_table2 = pd.crosstab(sessions_merged2["acquisition_channel"], sessions_merged2["converted"])
print(contingency_table2)

chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table2)

print(f"\nChi-square statistic: {chi2:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: Acquisition channel IS associated with conversion")
else:
    print("Result: Acquisition channel is NOT associated with conversion")

converted                0     1
acquisition_channel             
Email                 1708   277
email                11817  1771
organic              11079  1673
paid_Search           1206   160
paid_search          12003  1754
referral             11563  1746
social               11545  1798

Chi-square statistic: 6.7753
P-value: 0.3421
Result: Acquisition channel is NOT associated with conversion


Q6 Interpretation:
P-value: 0.3421 not significant.

How a customer was acquired has no association with whether they convert. All channels: email, social, referral, paid search and organic lead to similar conversion rates.

**Note: The data contains inconsistent capitalisation (e.g. "Email" and "email", "paid_Search" and "paid_search") which is a data quality issue that should be cleaned before drawing firm conclusions.**

Q7. Is product category associated with discount usage?

In [ ]:
contingency_table3 = pd.crosstab(transactions["product_category"], transactions["discount_used"])
print(contingency_table3)

chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table3)

print(f"\nChi-square statistic: {chi2:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: Product category IS associated with discount usage")
else:
    print("Result: Product category is NOT associated with discount usage")

discount_used        0     1
product_category            
beauty            5305  2281
electronics       5212  2212
fashion           5326  2255
home              5210  2278
sports            5225  2279
toys              5196  2221

Chi-square statistic: 1.4622
P-value: 0.9174
Result: Product category is NOT associated with discount usage


Q7 Interpretation:
Discounts are applied fairly evenly across all product categories around 2,200-2,280 discounted transactions per category.
P-value: 0.9174 not significant.

No product category attracts more discounts than others. Discounts are being applied randomly rather than strategically. A smarter approach would be to use discounts only on underperforming categories to boost their sales.

Q8. Is time on site correlated with purchase value?

In [ ]:
sessions_transactions = sessions.merge(transactions, on="customer_id")

corr, p_value = stats.pearsonr(sessions_transactions["time_on_site"], sessions_transactions["purchase_value"])

print(f"Correlation: {corr:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: There IS a statistically significant correlation")
else:
    print("Result: There is NO statistically significant correlation")

Correlation: -0.0017
P-value: 0.3479
Result: There is NO statistically significant correlation


Q8 Interpretation:
Correlation: -0.0017 — essentially zero.

P-value: 0.3479 not significant.

There is no relationship between time spent on site and how much someone spends. Keeping customers on the site longer does not lead to bigger purchases. Time on site is not a useful metric for predicting revenue.

Q9. Is number of pages viewed correlated with purchase value?

In [ ]:
corr, p_value = stats.pearsonr(sessions_transactions["pages_viewed"], sessions_transactions["purchase_value"])

print(f"Correlation: {corr:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: There IS a statistically significant correlation")
else:
    print("Result: There is NO statistically significant correlation")

Correlation: 0.0001
P-value: 0.9588
Result: There is NO statistically significant correlation


Q9 Interpretation:
Correlation: 0.0001  essentially zero.

P-value: 0.9588  not significant.

Pages viewed has no correlation with purchase value. Viewing more pages does not mean spending more money.

 *This is a contrast to Homework 1, where we found that viewing more pages was linked to higher conversion (more likely to buy). So pages viewed influences WHETHER someone buys, but not HOW MUCH they spend. These are two different things.*

Q10. Is time on site correlated with pages viewed?

In [ ]:
corr, p_value = stats.pearsonr(sessions["time_on_site"], sessions["pages_viewed"])

print(f"Correlation: {corr:.4f}")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: There IS a statistically significant correlation")
else:
    print("Result: There is NO statistically significant correlation")

Correlation: 0.0045
P-value: 0.2293
Result: There is NO statistically significant correlation


Q10 Interpretation:
Correlation: 0.0045 essentially zero.

P-value: 0.2293  not significant.

Time on site and pages viewed are completely independent of each other. Someone can spend a long time on one page without clicking anywhere, or quickly browse many pages in a short time. Neither behaviour predicts spending.

### Challenges

C1. Explore whether high-spending customers behave differently in terms of session time.

Define “high spending” yourself.

In [ ]:
# Define high spending as above the 75th percentile
threshold = merged["purchase_value"].quantile(0.75)
print(f"High spending threshold: ${threshold:.2f}")

high_spenders = merged[merged["purchase_value"] >= threshold]["customer_id"]
low_spenders = merged[merged["purchase_value"] < threshold]["customer_id"]

high_sessions = sessions[sessions["customer_id"].isin(high_spenders)]["time_on_site"]
low_sessions = sessions[sessions["customer_id"].isin(low_spenders)]["time_on_site"]

t_stat, p_value = stats.ttest_ind(high_sessions, low_sessions)

print(f"High spenders avg time on site: {high_sessions.mean():.2f} seconds")
print(f"Low spenders avg time on site: {low_sessions.mean():.2f} seconds")
print(f"P-value: {p_value:.4f}")

if p_value < 0.05:
    print("Result: High spenders DO behave differently in session time")
else:
    print("Result: High spenders do NOT behave differently in session time")

High spending threshold: $101.03
High spenders avg time on site: 300.84 seconds
Low spenders avg time on site: 300.89 seconds
P-value: 0.9544
Result: High spenders do NOT behave differently in session time


C1 Interpretation:
I defined high spending as purchases above the 75th percentile  $101.03.

High spenders averaged 300.84 seconds on site vs 300.89 seconds for low spenders.

P-value: 0.9544  not significant.

High spenders and low spenders behave identically on the website. We cannot identify a high-value customer by how long they browse. This means the business cannot use session time to predict or target high spenders.

C2. Find a relationship (difference or correlation) that you believe is meaningful for business strategy.

Explain why.

In [ ]:
# C2: Discount usage vs purchase value - most meaningful business finding
discount_rate = merged["discount_used"].mean() * 100
revenue_lost = (merged[merged["discount_used"]==1]["purchase_value"].count()) * \
               (merged[merged["discount_used"]==0]["purchase_value"].mean() - \
                merged[merged["discount_used"]==1]["purchase_value"].mean())

print(f"Percentage of transactions using discounts: {discount_rate:.1f}%")
print(f"Average spend WITHOUT discount: ${merged[merged['discount_used']==0]['purchase_value'].mean():.2f}")
print(f"Average spend WITH discount: ${merged[merged['discount_used']==1]['purchase_value'].mean():.2f}")
print(f"Difference per transaction: ${merged[merged['discount_used']==0]['purchase_value'].mean() - merged[merged['discount_used']==1]['purchase_value'].mean():.2f}")
print(f"Estimated revenue impact of discounts: ${revenue_lost:,.2f}")

Percentage of transactions using discounts: 30.1%
Average spend WITHOUT discount: $80.76
Average spend WITH discount: $64.40
Difference per transaction: $16.37
Estimated revenue impact of discounts: $221,369.68


C2 Interpretation:
The most meaningful business finding in this entire dataset is the discount impact.

30.1% of all transactions use a discount, and those transactions generate $16.37 less each.

Multiplied across all discounted transactions, the estimated revenue impact is $221,369.
This is statistically confirmed with a p-value of 0.0000.

My recommendation: the business should audit its discount strategy immediately. Discounts should be restricted to specific underperforming categories or used as loyalty rewards, rather than being broadly available to all customers at all times.

c3. Identify a statistically significant result that may not be practically meaningful.

Explain why statistical significance ≠ business importance.

In [ ]:
# C3: Statistically significant but not practically meaningful
# Example: Compare purchase values across all countries
from scipy.stats import f_oneway

groups = [merged[merged["country"] == c]["purchase_value"]
          for c in merged["country"].unique()]

f_stat, p_value = f_oneway(*groups)

country_avgs = merged.groupby("country")["purchase_value"].mean()
print("Average purchase by country:")
print(country_avgs.round(2))
print(f"\nF-statistic: {f_stat:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Range of averages: ${country_avgs.max() - country_avgs.min():.2f}")

Average purchase by country:
country
Australia      75.55
Canada         76.72
France         75.75
Germany        75.76
Netherlands    75.51
Spain          75.63
UK             75.63
US             76.18
Name: purchase_value, dtype: float64

F-statistic: 0.2775
P-value: 0.9630
Range of averages: $1.21


C3: Statistical significance vs practical importance

The discount analysis (Q2) is a strong example of this concept.

The t-test gave a p-value of 0.0000, meaning the $16.37 difference
between discounted and non-discounted purchases is 100% statistically
real, not random.

However, we need to ask: is $16.37 practically meaningful enough to
change business strategy?

Arguments that it IS practically meaningful:
- Multiplied across 13,500 discounted transactions = $221,000 impact
- That represents real lost revenue at scale

Arguments that it may NOT be practically meaningful:
- The discount may be attracting customers who would not have purchased
  at full price at all — meaning some revenue is better than none
- We don't know if removing discounts would cause those customers to
  simply not buy, which would lose the 64 entirely rather than gaining
  the extra $16
- Statistical significance just tells us the difference is real
  it doesn't tell us what caused it or what would happen if we removed discounts

This is why statistical significance alone is never enough for a
business decision. You need context, causation, and business judgement
alongside the numbers.